### PR TITLE
#419: Add rule L037 -- warn if "ORDER BY" has a mixture of columns with explicit and default ordering

### PR DESCRIPTION
**Pull request by barrywhart**
Addresses #419

### CODE REVIEW

**Comment by barrywhart** [comment]
@NiallRees, @pwildenhain, @alanmcruickshank: I have the linting part of this PR working correctly, but for the fixes, I found a possible gap related to `LintFix`: `LintFix` requires the fix to *precede* some existing `BaseSegment`, but here I need to add new text at the very **end** of the query.

Before fix:
```
SELECT * FROM t ORDER BY a DESC, b
```

After fix:
```
SELECT * FROM t ORDER BY a DESC, b ASC
```

Is this supported? If not, how might we go about supporting it?

As an experiment, I tried passing in the parent element, but that placed the fix in the wrong place:
```
SELECT * FROM t ASC ORDER BY a DESC, b
```

**Comment by pwildenhain** [comment]
Responding quickly on my phone, please excuse the brevity and/or typos.

When I added the fixes for L019 I needed to place leading commas _before_ the previous segment and it was tricky.

I ended up using the a combination of switches that I passed through the `memory` parameter of the `LintResult` in order to trigger the different fix routines.

Does that give you any ideas? 💡

**Comment by barrywhart** [comment]
Thanks, @pwildenhain! My problem turned out to be simpler than I feared. It was a combination of a tricky "end of loop" condition and needing to combine two individual `LintFix` objects into a single, compound `LintFix`, in order to place new whitespace correctly.

File: src/sqlfluff/core/rules/std.py [hunk 1]
```diff
@@ -368,9 +369,7 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
         # Is this line just comments?
         if all(
             seg.is_type(
-                "whitespace",
-                "comment",
-                "indent",  # dedent is a subtype of indent
+                "whitespace", "comment", "indent"  # dedent is a subtype of indent
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std.py:372]: Unrelated changes from Black

File: src/sqlfluff/core/rules/std.py [hunk 2]
```diff
@@ -3407,6 +3399,126 @@ def _eval(self, segment, **kwargs):
...
+class OrderByColumnInfo(NamedTuple):
+    """For L035, segment that ends an ORDER BY column and any order provided."""
+
+    separator: BaseSegment
+    order: Optional[str]  # One of 'ASC'/'DESC'/None
+
+
+@std_rule_set.document_fix_compatible
+@std_rule_set.register
+class Rule_L035(BaseCrawler):
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std.py:3411]: I named this `Rule_L035` because I noticed there is a 34 and a 36, but no 35. Is this okay?

**Comment by dmateusp** [comment]
@barrywhart fyi I had another PR with rule ID L035 https://github.com/sqlfluff/sqlfluff/pull/577, and I think that raises the question of https://github.com/sqlfluff/sqlfluff/issues/680

we will need to resolve the conflict by renaming one of our rules to L037 but I think that could be done right before merging when one PR is approved

File: src/sqlfluff/core/rules/std.py (hunk 2 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/std.py:3411]: As per #680 can we rename to 37?

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Apart from renaming as per #680, this looks good.

**Comment by barrywhart** [comment]
I'll rename the rule and move to the bottom of the file before merging.

File: src/sqlfluff/core/rules/std.py (hunk 2 above)
Comment (barrywhart) [reply src/sqlfluff/core/rules/std.py:3411]: Will do!

**Comment by barrywhart** [comment]
@alanmcruickshank: Ok, the rule is renamed to 037 now. 🎉
